### PR TITLE
Re-implement vocabulary association mechanisms as an algorithm

### LINKED ISSUE #2378: Normative statements for the vocabulary association mechanism?
The [§11 Vocabulary Association Mechanism](https://w3c.github.io/epub-specs/epub33/rs/#sec-vocab-assoc) section is altogether a _MAY_; it starts with:

> [Reading systems](https://www.w3.org/TR/epub-33/#dfn-epub-reading-system) MAY support [vocabulary association mechanisms](https://www.w3.org/TR/epub-33/#sec-vocab-assoc) [[epub-33](https://w3c.github.io/epub-specs/epub33/rs/#bib-epub-33)] in [EPUB content documents](https://www.w3.org/TR/epub-33/#dfn-epub-content-document).

We actually know that RS-s almost never implement this. It is therefore a bit disturbing to see the section full of _MUST_ statements (which all would also require tests). It all sounds very "aspirational" to me. Wouldn't it be better to rephrase this section? Maybe a simple solution to replace all the _MUST_ statements by plain, lower case "must"-s?

We had a similar situation with, e.g., MOL, which is also an optional feature, but the section contains lots of _MUST_-s. But this section is not optional in terms of _SHOULD_, but in terms of _MAY_. Also, we know that MOL is implemented by RS-s (although not by all).

### LINKED ISSUE #2382: Clarify that the scheme attribute does not have a default vocabulary.
It's the only attribute that requires a prefix for its values.

### PR DESCRIPTION
**Pull request by mattgarrish**
Not looking for approval at this time. Just curious for early feedback on whether this makes sense as a replacement.

<strike>I kept the one requirement not to blow up when encountering unknown prefixes.</strike>

Fixes #2378 
Fixes #2382 

* [Preview](https://cdn.statically.io/gh/w3c/epub-specs/experiment/issue-2378/epub33/rs/index.html)
* [Diff](https://services.w3.org/htmldiff?doc1=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://w3c.github.io/epub-specs/epub33/rs/index.html&doc2=https://labs.w3.org/spec-generator/%3Ftype=respec%26url=https://cdn.statically.io/gh/w3c/epub-specs/experiment/issue-2378/epub33/rs/index.html)

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/pull/2379.html" title="Last updated on Aug 6, 2022, 12:11 PM UTC (30ed27d)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/epub-specs/2379/706c5a7...30ed27d.html" title="Last updated on Aug 6, 2022, 12:11 PM UTC (30ed27d)">Diff</a>

### CODE REVIEW

**Comment by iherman** [comment]
I think that is great. It would get around the issue of the awkward _MUST_ statements and it works well.

Thanks!

**Comment by mattgarrish** [comment]
Okay, I think this is good to go now. The last commit mostly added explanations for all the steps. I also made a few tweaks to infra-ize things a bit more.

The two changes of note are:

- I've added an optional item to the epub:type processing list to get an expanded url. It was a bit strange to describe processing the attribute without mentioning that you could get an expanded value to process.
- I also added a paragraph to the top of the algorithm clarifying that it's not required to expand property values to use them. Reading systems that don't support the vocabulary association mechanisms can just treat them like strings.

**Comment by mattgarrish** [comment]
And as reading the diff is a bit of a nightmare, here's a link to the current section if you want to manually compare against the preview: https://www.w3.org/TR/epub-rs-33/#sec-vocab-assoc

File: epub33/rs/index.html
Comment (iherman) [inline epub33/rs/index.html:2097]: I am not sure why, but on my cloned version of the document these references do not appear as references, just plain texts. This may be some respec/github black art... :-( If it works on your side, ignore this.

File: epub33/rs/index.html
Comment (iherman) [inline epub33/rs/index.html]:
```suggestion
									this case, the value is drawn from the default vocabulary as described in the next step of the algorithm.</p>
```

File: epub33/rs/index.html
Comment (iherman) [inline epub33/rs/index.html]: Here again I do not see the infra reference in my local copy... :-(

File: epub33/rs/index.html
Comment (iherman) [inline epub33/rs/index.html]: Is that specified somewhere? I thought a colon is an acceptable character in value... In which case the previous step should be modified to say that we split at the first occurrence of a colon...

**Review by iherman — APPROVED** [review]

**Comment by iherman** [comment]
I have sketched out an implementation in typescript, just to check. Emphasis is on "sketch", but it helped me to follow the infra stuff:-)

File: epub33/rs/index.html
Comment (mattgarrish) [reply epub33/rs/index.html:2097]: I'm getting proper links in my local version. Are all references broken for you?

File: epub33/rs/index.html
Comment (mattgarrish) [reply epub33/rs/index.html]: Must be a glitch either locally for you or that's been resolved now.

File: epub33/rs/index.html
Comment (mattgarrish) [reply epub33/rs/index.html]: Ya, I over-interpreted the relative path requirement a bit, but it looks to be more complicated than either of us has right now.

According to the [EBNF for the data type](https://w3c.github.io/epub-specs/epub33/core/#sec-property-datatype), the value has to be a [path-relative-scheme-less-URL string](https://url.spec.whatwg.org/#path-relative-scheme-less-url-string).

It says in that definition that the value "does not start with: a [URL-scheme string](https://url.spec.whatwg.org/#url-scheme-string), followed by U+003A (:)".

So if you circle down through the many definitions you get:

- A path-relative-scheme-less URL must be a path-relative-URL string
- A path-relative-URL string must be zero or more URL-path-segment strings, separated from each other by U+002F (/), and not start with U+002F (/)
- A URL-path-segment string must be zero or more URL units excluding U+002F (/) and U+003F (?)

So wouldn't this make a colon illegal in the first URL-path-segment string? If you have one, it has to be after the first slash.

Do we want to try and spec that out, though, or can we leave the validity to the end? The ultimate arbiter is that you get a valid URL string. Multiple colons is a pretty niche case.

Comment (mattgarrish) [reply epub33/rs/index.html]: Maybe we could test this after splitting. So, if there are one or more colons, split on the first. Then before moving to the next step say if `propertyName` contains a colon before a first slash, the value is invalid?

Comment (iherman) [reply epub33/rs/index.html]: If I follow the definition for the property:

- path-relative-scheme-less-URL string -> cannot start with a URL-scheme string followed by a (:) but otherwise is a path-relative-URL string. 
- a URL-scheme is ASCII alpha, followed by ASCII alphanumeric or '+','-','.'
- path-relative-URL string allows for the ':' character.

I.e., the value indeed cannot be "abcd:efgh". But I believe it can be something like "abcd/efg:hij". E.g., I believe this is valid:

```
… prefix="test: http://www.example.org/"
… property="test:abcd/#efg:hij"
```

yielding `http://www.example.org/abcd/#efg:hij`.

This means, I believe, that we should be prepared for the case where the property contains more than one ':' character; meaning that the split must happen at the first occurrence of ':' and then we are done...

Comment (iherman) [reply epub33/rs/index.html]: Just realized a practical example, which we all love: epubcfi. If a property includes an epubcfi fragment, it may have a ':' character... :-)

Comment (mattgarrish) [reply epub33/rs/index.html]: The latest commit rewrites step 2 to account for multiple colons.

I couldn't use the infra splitting algorithm anymore, as that could lead to multiple chunks which would then have to be recombined. Seemed excessive. I have it now to split on the first colon and assign the strings before and after to the appropriate variables. I'm assuming splitting is widely enough known that we don't have to cite anything.

Anyway, if there are multiple colons, these changes should only allow valid cases to pass through - where the colon is not in the first segment of the name. I added a couple of examples to the definition to help explain this: "foo:bar/ba:z" is valid while "foo:bar:baz/qux" is not.

If you don't think I've got this right, though, let me know.

Comment (iherman) [reply epub33/rs/index.html]: I am not sure the restriction is right: I believe `foo:bar#ba:z` is also a valid case. Ie, (3) would create a false negative, wouldn't it? There are URI schemes that do not rely on slash character (DID, ISBN, mailto)

I would rather go down the line of

- remove current (3)
- split at the first colon i.e., what is now (4)
- before setting the value of _propertyName_ check whether the value begins with a [URL Scheme string](https://url.spec.whatwg.org/#url-scheme-string) followed by a colon. If true, it is an error and return null, otherwise we set the _propertyName_ and we are done.

Would that work?

Comment (mattgarrish) [reply epub33/rs/index.html]: > check whether the value begins with a [URL Scheme string](https://url.spec.whatwg.org/#url-scheme-string) followed by a colon

Maybe simplest of all would be to say that `propertyName` must be a valid path-relative-scheme-less-URL string, since that's where that exception comes from and will match up with the EBNF. I agree these details of what is valid are better left to the explanation.

Comment (mattgarrish) [reply epub33/rs/index.html]: I've redone step 2 to make a few changes.

- step one no longer says to continue processing at the next step. Even if the value doesn't have a colon, that doesn't mean it's a valid PRSLU string.
- the former third step is gone
- after processing, `propertyReference` is checked that it is not an empty string or an invalid PRSLU

I also renamed `propertyName` to `propertyReference` to match the EBNF naming of the components.

Comment (mattgarrish) [reply epub33/rs/index.html]: Hm, this still may not be right. Looking at the PRSLU definition, an empty string is valid. That means that a prefix-only property value is valid, which also matches the CURIE definition. For example, you could do: `<meta ... scheme="myScheme:"

I know we considered this invalid in the past, as made some weird hacks to create schemes that would have both a prefix and reference (marc:relators comes to mind). But I guess it's valid, so I should take that test out.

Comment (mattgarrish) [reply epub33/rs/index.html]: See also https://github.com/w3c/epub-specs/issues/2382. I've fixed this, too.

Comment (iherman) [reply epub33/rs/index.html]: Yes indeed, I know that in the RDFa/Turtle practice using a CURIE without reference (ie, re-using the prefix as a URL on its own right) is actually fairly frequent. But I thought you have some good reasons to disallow the trailing ':' case...

Otherwise, it looks good to me. Personally, I do not see any reason not to merge this and consider it done...

Comment (mattgarrish) [reply epub33/rs/index.html]: > But I thought you have some good reasons to disallow the trailing ':' case...

Nope... 😄 

The unfortunate truth is that we lacked the requisite expertise to be implementing RDFa in the package document when we tried to build this framework for 3.0. We probably should have abandoned it completely when it was pointed out we were absuing the `about` attribute, but rewriting the whole framework at the last minute likely wouldn't have had much better results.

In this case, I think the definition of an optional prefix followed by a reference is what confused people. Whoever made the first claim that it meant the reference couldn't be an empty string was just accepted as gospel. I know this is the deepest I've ever dug into the fine print, so for that alone this issue has been insightful.

But, yes, I think this captures everything we need now. Do you think we need a group resolution before merging tihs, since we haven't discussed doing this kind of a rewrite? It's pretty substantive, at least in terms of removing normative requirements.

Comment (mattgarrish) [reply epub33/rs/index.html]: I've added a final clarification to the data type definition that a prefix and colon with an empty reference is a valid value. I added this after the statement that an empty string is technically valid but non-conforming, as that's likely where the confusion arose from.

Comment (iherman) [reply epub33/rs/index.html]: Yes, it would be good to have a WG approval. Except that... @wareid will be out for the next two weeks on vacations, and she should not be worrying about the WG. I do not know when @dauwhe is back, not sure that he will be around next week. Also, there is no real reason these days to hold WG calls, it is all quiet with some people making tests...

Let us wait if @dauwhe is back next week. What we could do is that you send a mail to the WG explaining what has been done, and leave a week to respond/object to the changes. I would hate to leave this PR open for several weeks...

Comment (mattgarrish) [reply epub33/rs/index.html]: >  Also, there is no real reason these days to hold WG calls

Hm, ya, I wonder if I'm overthinking this. We're not adding new requirements, only loosening existing ones. And it's for an optional feature that we haven't heard has been implemented.

It's noted in the change log, too, so anyone keeping track of what's going on will find it. Maybe we should just go ahead an merge...

Comment (iherman) [reply epub33/rs/index.html]: I would like to get at least @dauwhe 's reaction (@wareid is out on vacations)

**Comment by iherman** [comment]
The issue was discussed in a [meeting](https://www.w3.org/publishing/groups/epub-wg/Meetings/Minutes/2022-08-04-epub#section2) on 2022-08-04

List of resolutions:

- ***[Resolution No. 1](https://www.w3.org/publishing/groups/epub-wg/Meetings/Minutes/2022-08-04-epub#resolution1):  The WG supports merging #2379 pending approval from the assigned reviewers..***

<details><summary><i>View the transcript</i></summary>

### 2. Re-implement vocabulary association mechanisms as an algorithm (pr epub-specs#2379)

_See github pull request [epub-specs#2379](https://github.com/w3c/epub-specs/pull/2379)._




_See github issue [epub-specs#2378](https://github.com/w3c/epub-specs/issues/2378)._




_See github issue [epub-specs#2382](https://github.com/w3c/epub-specs/issues/2382)._




**Dave Cramer:** there are a couple issues associated with this (#2378, #2382).  
… this is very technical, I wish mgarrish were here to explain.  
… doesn't really seems to change how the spec works, more an editorial change.  
… seems like they were just looking for WG approval before merging, even though most of us probably won't have strong opinions on this.  
… i'm generally in favor of explaining things in an algorithmic fashion rather than hard to understand paragraphs.  
… maybe we just say that we don't have objection to merging this pending github reviews?.  

> **Proposed resolution: The WG supports merging #2379 pending approval from the assigned reviewers..** *(Dave Cramer)*

> *Dave Cramer:* +1.

> *Matthew Chan:* +1.

> *Brady Duga:* +1.

> *Toshiaki Koike:* +1.

> *Shinya Takami (高見真也):* +1.

> ***Resolution #1: The WG supports merging #2379 pending approval from the assigned reviewers..***

</details>

**Comment by iherman** [comment]
Per WG resolution, we need an approval of at least @shiestyle and @dauwhe (@wareid and, I presume, @rdeltour being on vacations...). Shinya, Dave, could you do this, please?

**Comment by mattgarrish** [comment]
I think they gave us the green light to merge it once we are happy with it. I think the "reviewers" in this case means me and you, as it sound like the group is fine with whatever we resolve on for this given no one's hugely passionate about a feature that's only for internal reading system use... :smile:

**Pull request merged**
